File: problem/0094_Binary Tree Inorder Traversal.js
// Given the root of a binary tree, return the inorder traversal of its nodes' values.

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

// inOrder: node.left -> node -> node.right
var inorderTraversal = function(root) {
    const result = [];
    const inorder = (node) => {
        if(!node) return;
        let l = inorder(node.left)
        result.push(node.val)
        let r = inorder(node.right)
    };
    inorder(root);
    return result;
};

File: c#-solution/0094. Binary Tree Inorder Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
  public IList<int> InorderTraversal(TreeNode root)
  {
    List<int> ans = new List<int>();
    Stack<TreeNode> stack = new Stack<TreeNode>();
    while (stack.Count != 0 || root != null)
    {
      if (root != null)
      {
        stack.Push(root);
        root = root.left;
      }
      else
      {
        root = stack.Pop();
        ans.Add(root.val);
        root = root.right;
      }
    }
    return ans;
  }
}

// not clear solution
public class Solution {
  List<int> resultList = new List<int>();
  public IList<int> InorderTraversal(TreeNode root) {
    if(root == null) return new List<int>();

    InorderTraversal(root.left);
    resultList.Add(root.val);
    InorderTraversal(root.right);
    return resultList;
  }
}